### PR TITLE
Fix Issue 24010 - Destructor called before end of scope for tuples

### PR DESCRIPTION
**Pull request by BorisCarvajal**
Expand tuple declarations to separate statements, this way each variable gets the usual destruction/exception code attached.

### CODE REVIEW

**Comment by RazvanN7** [comment]
Please target stable.

cc @WalterBright as you were looking into the same issue.

**Review by dkorpel — APPROVED** [review]

File: compiler/src/dmd/statementsem.d [hunk 1]
```diff
@@ -279,6 +279,25 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
...
+            // expand tuple variables in order to attach destruction/exception logic
+            if (auto es = s.isExpStatement())
+            {
+                if (es.exp && es.exp.isDeclarationExp())
+                {
+                    auto de = es.exp.isDeclarationExp();
+                    auto vd = de.declaration.isVarDeclaration();
+                    if (vd && vd.aliasTuple && vd.aliasTuple.objects.length)
+                    {
+                        Statements* a = new Statements();
```
Comment (WalterBright) [inline compiler/src/dmd/statementsem.d:292]: This creates an unnecessary allocation. See https://github.com/dlang/dmd/pull/15366/files#diff-11fb0212c55c05ed6e743e6776abb3c56996381760c8725255c622f96f5d9d4bR308 instead.

File: compiler/test/runnable/test24010.d [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+void test2() {
+    bar!Foo();
+}
+/****************************/
+
+void main()
+{
+    test1();
+    test2();
```
Comment (WalterBright) [inline compiler/test/runnable/test24010.d:59]: The tests in https://github.com/dlang/dmd/pull/15366/files#diff-d6cdd89ae270292efd2462b4f14ab839114f4ec71ddb37d52eb9d84e3238965bR4838
1. verify the correct order of the destructor calls
2. verify that the tuple name can also be used to access the fields
3. is integrated into existing sdtor.d tests, where the other destructor tests are, and merging into existing test files helps the test suite run faster

So, please replace this test with the other one.

**Review by WalterBright — DISMISSED** [review]
Excellent work!

File: compiler/src/dmd/statementsem.d (hunk 1 above)
Comment (BorisCarvajal) [reply compiler/src/dmd/statementsem.d]: Done, even reduced to just one `insert`. (thanks to #15369)

File: compiler/test/runnable/test24010.d (hunk 2 above)
Comment (BorisCarvajal) [reply compiler/test/runnable/test24010.d]: Done.

**Comment by BorisCarvajal** [comment]
Currently targeting master, but if stable is required, we need to cherry pick #15369.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**